File: te_phoenix/lib/te_phoenix/battle/morale.ex
defmodule TePhoenix.Battle.Morale do
  @moduledoc """
  Enemy morale system — fear, flee AI, rout.
  Ported from te/battle/morale.js

  Morale (0-100) drops from ally deaths, crits, heavy damage, low HP.
  Rises from kills, heals, idle turns.
  Below flee threshold: AI attempts to flee.
  Leader flees: all allies rout.
  """

  alias TePhoenix.Battle.Combatant

  @doc "Initialize morale for all combatants in a battle"
  def init_morale(state) do
    settings = state.settings
    if not settings[:enable_morale], do: state

    starting = settings[:starting_morale] || 100
    max_morale = settings[:max_morale] || 100
    brave_bonus = settings[:personality_brave_bonus] || 20
    coward_penalty = settings[:personality_coward_penalty] || -20

    combatants =
      Enum.reduce(state.combatants, state.combatants, fn {id, c}, combs ->
        personality = Map.get(c, :personality, "") |> to_string() |> String.downcase()

        bonus = case personality do
          "brave" -> brave_bonus
          "coward" -> coward_penalty
          _ -> 0
        end

        morale = min(max_morale, starting + bonus)
        is_fanatic = personality == "fanatic"
        is_leader = Map.get(c, :is_boss, false)

        c = c
        |> Map.put(:morale, morale)
        |> Map.put(:morale_fanatic, is_fanatic)
        |> Map.put(:morale_is_leader, is_leader)

        Map.put(combs, id, c)
      end)

    %{state | combatants: combatants}
  end

  @doc "Adjust morale for a combatant (AI only, fanatics immune)"
  def adjust_morale(state, combatant_id, delta) do
    settings = state.settings
    if not settings[:enable_morale], do: state

    case Map.get(state.combatants, combatant_id) do
      nil -> state
      c ->
        if not c.is_ai or Map.get(c, :morale_fanatic, false) do
          state
        else
          max_m = settings[:max_morale] || 100
          new_morale = max(0, min(max_m, (c.morale || 100) + delta))
          c = %{c | morale: new_morale}
          %{state | combatants: Map.put(state.combatants, combatant_id, c)}
        end
    end
  end

  @doc "Handle morale changes when a combatant dies. Returns {state, flee_events}"
  def on_combatant_death(state, dead_id) do
    settings = state.settings
    if not settings[:enable_morale], do: {state, []}

    dead = Map.get(state.combatants, dead_id)
    if dead == nil, do: {state, []}

    dead_team = dead.team_id

    # Morale boost for enemies
    enemy_gain = settings[:enemy_kill_gain] || 10
    state =
      state.teams
      |> Enum.reject(fn {tid, _} -> tid == dead_team end)
      |> Enum.reduce(state, fn {_tid, members}, s ->
        Enum.reduce(members, s, fn id, s2 -> adjust_morale(s2, id, enemy_gain) end)
      end)

    # Morale loss for allies
    is_leader = Map.get(dead, :morale_is_leader, false)
    loss = if is_leader, do: settings[:leader_death_loss] || 30, else: settings[:ally_death_loss] || 20

    {state, flee_events} =
      (Map.get(state.teams, dead_team, []) -- [dead_id])
      |> Enum.reduce({state, []}, fn id, {s, events} ->
        s = adjust_morale(s, id, -loss)
        c = Map.get(s.combatants, id)

        if c && should_flee?(s, id) do
          reason = if is_leader, do: :leader_fell, else: :morale_broken
          {s, [%{type: :morale_flee, combatant_id: id, name: c.name, reason: reason} | events]}
        else
          {s, events}
        end
      end)

    {state, Enum.reverse(flee_events)}
  end

  @doc "Handle morale changes when a combatant takes damage"
  def on_damage_taken(state, target_id, damage, is_crit) do
    settings = state.settings
    if not settings[:enable_morale], do: state

    state = if is_crit do
      adjust_morale(state, target_id, -(settings[:critical_hit_loss] || 10))
    else
      state
    end

    target = Map.get(state.combatants, target_id)
    if target == nil, do: state

    dmg_pct = damage / max(1, target.max_hp)
    state = if dmg_pct >= (settings[:heavy_damage_pct] || 0.30) do
      adjust_morale(state, target_id, -(settings[:heavy_damage_loss] || 10))
    else
      state
    end

    # Low HP penalty (once per combatant)
    hp_pct = target.current_hp / max(1, target.max_hp)
    if hp_pct < (settings[:low_hp_threshold] || 0.25) and not Map.get(target, :low_hp_morale_applied, false) do
      target = Map.put(target, :low_hp_morale_applied, true)
      state = %{state | combatants: Map.put(state.combatants, target_id, target)}
      adjust_morale(state, target_id, -(settings[:low_hp_loss] || 15))
    else
      state
    end
  end

  @doc "Morale recovery when healed"
  def on_heal_received(state, target_id) do
    gain = state.settings[:heal_received_gain] || 5
    adjust_morale(state, target_id, gain)
  end

  @doc "Morale recovery on idle turn (no damage taken)"
  def on_idle_turn(state, combatant_id) do
    gain = state.settings[:idle_turn_gain] || 3
    adjust_morale(state, combatant_id, gain)
  end

  @doc "Check if a combatant should flee (morale below threshold)"
  def should_flee?(state, combatant_id) do
    settings = state.settings
    if not settings[:enable_morale], do: false

    case Map.get(state.combatants, combatant_id) do
      nil -> false
      c ->
        if not c.is_ai or Map.get(c, :morale_fanatic, false) do
          false
        else
          (c.morale || 100) <= (settings[:flee_threshold] || 20)
        end
    end
  end

  @doc "Rout: if a fleeing combatant is a leader, force all allies to flee"
  def check_rout(state, fleeing_id) do
    settings = state.settings
    if not settings[:enable_morale] or not (settings[:rout_on_leader_flee] != false), do: {state, []}

    fleeing = Map.get(state.combatants, fleeing_id)
    if fleeing == nil or not Map.get(fleeing, :morale_is_leader, false), do: {state, []}

    team = fleeing.team_id
    members = Map.get(state.teams, team, []) -- [fleeing_id]

    {state, routed} =
      Enum.reduce(members, {state, []}, fn id, {s, events} ->
        c = Map.get(s.combatants, id)
        if c && Combatant.alive?(c) and not Map.get(c, :morale_fanatic, false) do
          c = %{c | morale: 0}
          s = %{s | combatants: Map.put(s.combatants, id, c)}
          {s, [%{type: :rout, combatant_id: id, name: c.name} | events]}
        else
          {s, events}
        end
      end)

    {state, Enum.reverse(routed)}
  end

  @doc "Pursuit bonus damage against fleeing enemies"
  def pursuit_bonus(state, target_id) do
    if should_flee?(state, target_id) do
      state.settings[:pursuit_bonus_damage_pct] || 0.50
    else
      0
    end
  end
end

File: te.retired/battle/morale.js
// =================================================================
// ENEMY MORALE SYSTEM — Fear, Flee AI, Rout
// =================================================================
// ARCHITECTURE:
//   Every combatant has a morale value (0-100, starts at 100).
//   Morale drops when:
//     - An ally dies (-20)
//     - The combatant takes a critical hit (-10)
//     - HP drops below 25% (-15)
//     - A powerful attack lands (damage > 30% max HP: -10)
//     - The leader/boss dies (-30 to all allies)
//   Morale rises when:
//     - An enemy is killed (+10)
//     - A heal is received (+5)
//     - Turn passes without taking damage (+3)
//
//   Below the flee threshold (default 20), enemies attempt to flee.
//   If the leader flees, all allies rout (forced flee).
//   Players can pursue fleeing enemies for bonus attacks.
//
// DATA-DRIVEN:
//   All thresholds configurable via battle settings.
//   AI personality (brave/coward/fanatic) modifies behavior.
// =================================================================

const MORALE_DEFAULTS = {
    starting_morale: 100,
    max_morale: 100,
    flee_threshold: 20,
    // Morale loss triggers
    ally_death_loss: 20,
    leader_death_loss: 30,
    critical_hit_loss: 10,
    heavy_damage_loss: 10,
    heavy_damage_pct: 0.30,
    low_hp_loss: 15,
    low_hp_threshold: 0.25,
    // Morale gain triggers
    enemy_kill_gain: 10,
    heal_received_gain: 5,
    idle_turn_gain: 3,
    // Personality modifiers
    personality_brave_bonus: 20,
    personality_coward_penalty: -20,
    personality_fanatic_immune: true,
    // Pursuit
    pursuit_bonus_damage_pct: 0.50,
    // Rout
    rout_on_leader_flee: true,
};

function initMorale(battle) {
    if (!battle._settings?.enable_morale) return;
    for (const c of Object.values(battle.combatants)) {
        const base = battle._settings.starting_morale ?? MORALE_DEFAULTS.starting_morale;
        const personality = (c.personality || c.ai_personality || '').toLowerCase();
        let bonus = 0;
        if (personality === 'brave') bonus = battle._settings.personality_brave_bonus ?? MORALE_DEFAULTS.personality_brave_bonus;
        if (personality === 'coward') bonus = battle._settings.personality_coward_penalty ?? MORALE_DEFAULTS.personality_coward_penalty;
        c._morale = Math.min(battle._settings.max_morale ?? MORALE_DEFAULTS.max_morale, base + bonus);
        c._moraleFanatic = personality === 'fanatic';
        c._moraleIsLeader = c.is_boss || c.isBoss || false;
    }
}

function adjustMorale(battle, combatantId, delta, reason) {
    if (!battle._settings?.enable_morale) return;
    const c = battle.combatants[combatantId];
    if (!c || !c.isAI) return; // only AI has morale
    if (c._moraleFanatic) return; // fanatics never lose morale
    const max = battle._settings.max_morale ?? MORALE_DEFAULTS.max_morale;
    c._morale = Math.max(0, Math.min(max, (c._morale || 100) + delta));
}

// Called when a combatant dies
function onCombatantDeath(battle, deadId) {
    if (!battle._settings?.enable_morale) return [];
    const dead = battle.combatants[deadId];
    if (!dead) return [];

    const deadTeam = battle.getTeamId(deadId);
    const events = [];

    // Morale boost for enemies
    for (const [tid, members] of Object.entries(battle.teams)) {
        if (tid === deadTeam) continue;
        for (const id of members) {
            adjustMorale(battle, id, battle._settings.enemy_kill_gain ?? MORALE_DEFAULTS.enemy_kill_gain, 'enemy_killed');
        }
    }

    // Morale loss for allies
    const isLeader = dead._moraleIsLeader;
    const loss = isLeader
        ? (battle._settings.leader_death_loss ?? MORALE_DEFAULTS.leader_death_loss)
        : (battle._settings.ally_death_loss ?? MORALE_DEFAULTS.ally_death_loss);

    for (const id of (battle.teams[deadTeam] || [])) {
        if (id === deadId) continue;
        adjustMorale(battle, id, -loss, isLeader ? 'leader_died' : 'ally_died');
        const c = battle.combatants[id];
        if (c && shouldFlee(battle, id)) {
            events.push({ type: 'morale_flee', combatantId: id, name: c.name, reason: isLeader ? 'leader_fell' : 'morale_broken' });
        }
    }

    return events;
}

// Called when a combatant takes damage
function onDamageTaken(battle, targetId, damage, isCrit) {
    if (!battle._settings?.enable_morale) return;
    const target = battle.combatants[targetId];
    if (!target) return;

    if (isCrit) {
        adjustMorale(battle, targetId, -(battle._settings.critical_hit_loss ?? MORALE_DEFAULTS.critical_hit_loss), 'critical_hit');
    }

    const dmgPct = damage / Math.max(1, target.maxHp);
    if (dmgPct >= (battle._settings.heavy_damage_pct ?? MORALE_DEFAULTS.heavy_damage_pct)) {
        adjustMorale(battle, targetId, -(battle._settings.heavy_damage_loss ?? MORALE_DEFAULTS.heavy_damage_loss), 'heavy_damage');
    }

    // Low HP penalty only triggers once per combatant (not every hit)
    const hpPct = target.currentHp / Math.max(1, target.maxHp);
    if (hpPct < (battle._settings.low_hp_threshold ?? MORALE_DEFAULTS.low_hp_threshold) && !target._lowHpMoralePenaltyApplied) {
        target._lowHpMoralePenaltyApplied = true;
        adjustMorale(battle, targetId, -(battle._settings.low_hp_loss ?? MORALE_DEFAULTS.low_hp_loss), 'low_hp');
    }
}

// Called when healed
function onHealReceived(battle, targetId) {
    if (!battle._settings?.enable_morale) return;
    adjustMorale(battle, targetId, battle._settings.heal_received_gain ?? MORALE_DEFAULTS.heal_received_gain, 'healed');
}

// Called at end of turn if combatant wasn't attacked
function onIdleTurn(battle, combatantId) {
    if (!battle._settings?.enable_morale) return;
    adjustMorale(battle, combatantId, battle._settings.idle_turn_gain ?? MORALE_DEFAULTS.idle_turn_gain, 'idle_recovery');
}

function shouldFlee(battle, combatantId) {
    if (!battle._settings?.enable_morale) return false;
    const c = battle.combatants[combatantId];
    if (!c || !c.isAI) return false;
    if (c._moraleFanatic) return false;
    return (c._morale || 100) <= (battle._settings.flee_threshold ?? MORALE_DEFAULTS.flee_threshold);
}

// Rout: if a fleeing combatant is a leader, force all allies to flee
function checkRout(battle, fleeingId) {
    if (!battle._settings?.enable_morale) return [];
    if (!(battle._settings.rout_on_leader_flee ?? MORALE_DEFAULTS.rout_on_leader_flee)) return [];
    const fleeing = battle.combatants[fleeingId];
    if (!fleeing || !fleeing._moraleIsLeader) return [];

    const team = battle.getTeamId(fleeingId);
    const routed = [];
    for (const id of (battle.teams[team] || [])) {
        if (id === fleeingId) continue;
        const c = battle.combatants[id];
        if (c && c.currentHp > 0 && !c._knockedOut && !c._moraleFanatic) {
            c._morale = 0;
            routed.push({ type: 'rout', combatantId: id, name: c.name });
        }
    }
    return routed;
}

// Pursuit: bonus damage to fleeing enemies
function getPursuitBonus(battle, targetId) {
    if (!battle._settings?.enable_morale) return 0;
    if (!shouldFlee(battle, targetId)) return 0;
    return battle._settings.pursuit_bonus_damage_pct ?? MORALE_DEFAULTS.pursuit_bonus_damage_pct;
}

function getMoraleState(battle) {
    if (!battle._settings?.enable_morale) return null;
    const state = {};
    for (const [id, c] of Object.entries(battle.combatants)) {
        if (c._morale !== undefined) {
            state[id] = {
                morale: c._morale,
                fleeing: shouldFlee(battle, id),
                fanatic: c._moraleFanatic || false,
                leader: c._moraleIsLeader || false,
            };
        }
    }
    return state;
}

module.exports = {
    MORALE_DEFAULTS,
    initMorale,
    adjustMorale,
    onCombatantDeath,
    onDamageTaken,
    onHealReceived,
    onIdleTurn,
    shouldFlee,
    checkRout,
    getPursuitBonus,
    getMoraleState,
};
